py/tests/integration/test_spend_tx.py

```diff
@@ -88,6 +88,83 @@ def test_successful():
     common.stop_node(alice_node)
     shutil.rmtree(root_dir)
 
+def test_not_enough_tokens():
+    # Bob should not be able to send more tokens than he has
+    #
+    # Let's say Bob has 100 tokens. He should not be able to send more than
+    # 100 tokens to Alice.
+    #
+    # If there's an incoming but unconfirmed deposit into Bob's account then Bob
+    # should not be able to use the incoming tokens until the spend transaction
+    # they are in is confirmed.
+    test_settings = settings["test_not_enough_tokens"]
+    coinbase_reward = common.coinbase_reward() 
+    (root_dir, bob_node, bob_api, bob_top) = setup_node_with_tokens(test_settings, "bob") 
+
+    spend_tx_amt = test_settings["spend_tx"]["amount"]
+    spend_tx_fee = test_settings["spend_tx"]["fee"]
+    bob_balance = bob_api.get_account_balance()
+    bob_has_not_enough_tokens = bob_balance.balance < spend_tx_amt + spend_tx_fee
+    assert_equals(bob_has_not_enough_tokens, True)
+    print("Bob initial balance is " + str(bob_balance.balance) +
+            " and he will try to spend " + str(spend_tx_amt + spend_tx_fee))
+
+    alice_balance0 = {"balance":0} 
+    try:
+        alice_balance0 = bob_api.get_account_balance(pub_key=test_settings["spend_tx"]["alice_pubkey"])
+    except ApiException as e:
+        assert_equals(e.status, 404) # Alice has no account yet
+
+    # Bob tries to send some tokens to Alice
+    bob_internal_api = common.internal_api(bob_node)
+    spend_tx_obj = SpendTx(
+        recipient_pubkey=test_settings["spend_tx"]["alice_pubkey"],
+        amount=spend_tx_amt,
+        fee=spend_tx_fee)
+    print("Bob's spend_tx is " + str(spend_tx_obj))
+    bob_internal_api.post_spend_tx(spend_tx_obj)
+    print("Transaction sent")
+
+    # ensure Bob balance had not changed
+    bob_balance1 = bob_api.get_account_balance()
+    assert_equals(bob_balance.balance, bob_balance1.balance)
+
+    # ensure Alice balance had not changed
+    try:
+        alice_balance1 = bob_api.get_account_balance(pub_key=test_settings["spend_tx"]["alice_pubkey"])
+        assert_equals(alice_balance1.balance, alice_balance0.balance)
+    except ApiException as e:
+        assert_equals(e.status, 404) # Alice has no account yet
+
+    # wait for a block to be mined
+    print("Waiting for a next block to be mined")
+    common.wait_until_height(bob_api, bob_top.height + 1)
+
+    bob_new_top = bob_api.get_top()
+    bob_new_balance = bob_api.get_account_balance()
+    
+    blocks_mined = bob_new_top.height - bob_top.height
+
+    # Since Bob had mined the block he is receiving the coinbase_reward
+    expected_balance = bob_balance.balance + coinbase_reward * blocks_mined
+    assert_equals(bob_new_balance.balance, expected_balance)
```

Comment:
This assertion is checking that, even if further blocks have been mined by Bob after the spend tx has been generated, the balance of Bob was not charged with the spend tx. Though what is not clear to me is how does the test ensures that the spend tx was considered for inclusion in block in the first place:
* The [line generating the spend tx](https://github.com/aeternity/epoch/pull/383/files#diff-4bd03bf9816115cc882755103452746eR125) does not assert success - or does it by simply no exception raised? Maybe an assert on the status code would clarify?
* Also, even asserting that spend tx was generated does not clarify whether tx was considered for inclusion in block. What is the expected result of a spend tx issued with not enough amount in the account? Is the spend tx meant to be included in a block and the fee taken? (Maybe a distinct mining node from Bob is needed for that?)

Please comment.

Reply:
What if the fee is more than account's balance?
Currently if account does not have enough tokens for the transaction - the transaction is [ignored](https://github.com/aeternity/epoch/blob/master/apps/aecore/src/txs/aec_tx.erl#L125-L128) (not being included in the block) and no fee is applied for it.
Also [there is not much to be checked in the return code](https://github.com/aeternity/epoch/blob/master/apps/aehttp/src/aehttp_dispatch_int.erl#L11) since it doesn't validate that there are enough tokens to execute the tx